app/code/community/FrontCommerce/Payline/Helper/Duplication.php

```diff
@@ -0,0 +1,234 @@
+<?php
+
+/**
+ * This helper contains the interactions duplicated and adapted
+ * from the Payline module so that they can be used in a headless context
+ */
+class FrontCommerce_Payline_Helper_Duplication
+{
+  /**
+   * Source: Monext_Payline_Helper_Widget::getDataToken()
+   */
+  public function getDataToken($quote)
+  {
+    try {
+      $orderId = $this->_getReservedOrderId($quote);
+      $quote->setRealOrderId($orderId);
+
+      $array = Mage::helper('payline/payment')->initWithQuote($quote);
+      $array['version'] = Monext_Payline_Helper_Data::VERSION;
+      $array['payment']['action'] = Mage::getStoreConfig('payment/PaylineCPT/payline_payment_action');
+      $array['payment']['mode'] = 'CPT';
+
+      $returnUrl = Mage::getUrl('payline/index/cptReturnWidget');
+      $array['payment']['contractNumber'] = $this->getDefaultContractNumberForWidget();
+      $array['contracts'] = $this->getContractsForWidget(true);
+      $array['secondContracts'] = $this->getContractsForWidget(false);
+      if (empty($array['secondContracts'])) {
+        $array['secondContracts'] = array('');
+      }
+
+      // if ($forShortcut) {
+      //   $returnUrl = Mage::getUrl('payline/index/cptWidgetShortcut');
+      //   $array['payment']['contractNumber'] = $this->getContractNumberForWidgetShortcut();
+      //   $array['contracts'] = array($array['payment']['contractNumber']);
+      //   //                    $array['secondContracts'] = array('');
+      // }
+
+
+      $paylineSDK = Mage::helper('payline')->initPayline('CPT', $array['payment']['currency']);
+      // ToDo: use FC url helper
+      Mage::log('url: ' . $returnUrl, null, 'debug.log');
+      $paylineSDK->returnURL          = 'http://www.front-commerce.test/default/payline/process/widget/payline/process/widget'; //$returnUrl;
```

Comment:
Won't this need to be updated before shipping? Or maybe is this URL not really used in a SPA context?

Also I'm surprised that this URL has `payline/process/widget` twice.

Reply:
![](https://media.giphy.com/media/fnLwSQ4U5oahz5A7Ju/giphy-downsized.gif)

Oops! Thanks for your feebdack. It was indeed not aimed at being there :)

I amended the commit.